using variables in unsafe block to use sort() on an array

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: windows, Microsoft Windows 10 Pro v19042 64-bit
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz
CC version: Error: exec failed (CreateProcess) with code 2: The system cannot find the file specified.

 cmd: cc --version

getwd: C:\Users\User\Documents\v\rep_error
vmodules: C:\Users\User\.vmodules
vroot: C:\Users\User\Documents\v_store\v
vexe: C:\Users\User\Documents\v_store\v\v.exe
vexe mtime: 2021-11-03 21:30:47
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 dcf230c.2b41549

Git version: git version 2.33.1.windows.1
Git vroot status: weekly.2021.44-7-g2b415491 (6 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-windows-amd64 0f5168cd
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg .\founderror.v`
```v

fn main() {
	array:= [[1, 5], [4, 2]]
	println(array)
	i:= 0
	unsafe{array.sort(a[0]>b[0])}
	println(array)
	unsafe{array.sort(a[i]<b[i])} // issue here
	println(array)

}

```

**What did you expect to see?**

either [[1,5],[4,2]] or a error message saying "you cant use a variable here"(or something).


**What did you see instead?**
```
==================
C:/Users/User/AppData/Local/Temp/v/founderror.10995277303334667339.tmp.c:442: warning: WINVER redefined
C:/Users/User/AppData/Local/Temp/v/founderror.10995277303334667339.tmp.c:1884: error: 'i' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang```

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -3296,6 +3296,7 @@ fn (mut c Checker) array_builtin_method_call(mut node ast.CallExpr, left_type as
 			if mut node.args[0].expr is ast.LambdaExpr {
 				c.support_lambda_expr_in_sort(elem_typ.ref(), ast.bool_type, mut node.args[0].expr)
 			} else if node.args[0].expr is ast.InfixExpr {
+				c.check_sort_external_variable_access(node.args[0].expr)
 				if node.args[0].expr.op !in [.gt, .lt] {
 					c.error('`.${method_name}()` can only use `<` or `>` comparison',
 						node.pos)

--- a/vlib/v/checker/infix.v
+++ b/vlib/v/checker/infix.v
@@ -1071,3 +1071,40 @@ fn (mut c Checker) autocast_in_if_conds(mut right ast.Expr, from_expr ast.Expr, 
 		else {}
 	}
 }
+
+fn (mut c Checker) check_sort_external_variable_access(node ast.Expr) bool {
+	match node {
+		ast.InfixExpr {
+			if !c.check_sort_external_variable_access(node.left) {
+				return false
+			}
+			if !c.check_sort_external_variable_access(node.right) {
+				return false
+			}
+		}
+		ast.Ident {
+			if node.name !in ['a', 'b'] {
+				c.error('can not access external variable `${node.name}`', node.pos)
+				return false
+			}
+		}
+		ast.CallExpr {
+			return c.check_sort_external_variable_access(node.left)
+		}
+		ast.SelectorExpr {
+			return c.check_sort_external_variable_access(node.expr)
+		}
+		ast.IndexExpr {
+			if !c.check_sort_external_variable_access(node.left) {
+				return false
+			}
+			if !c.check_sort_external_variable_access(node.index) {
+				return false
+			}
+		}
+		else {
+			return true
+		}
+	}
+	return true
+}

--- a/vlib/v/checker/tests/array_sort_err.out
+++ b/vlib/v/checker/tests/array_sort_err.out
@@ -19,6 +19,12 @@ vlib/v/checker/tests/array_sort_err.vv:5:6: error: `.sort()` cannot use same arg
       |         ~~~~~~~~~~~
     6 |     arr.sort(c > d)
     7 | }
+vlib/v/checker/tests/array_sort_err.vv:6:11: error: can not access external variable `c`
+    4 |     arr.sort(a == b)
+    5 |     arr.sort(a > a)
+    6 |     arr.sort(c > d)
+      |              ^
+    7 | }
 vlib/v/checker/tests/array_sort_err.vv:6:6: error: `.sort()` can only use `a` or `b` as argument, e.g. `arr.sort(a < b)`
     4 |     arr.sort(a == b)
     5 |     arr.sort(a > a)

--- a/vlib/v/checker/tests/qsort_external_variable_err.out
+++ b/vlib/v/checker/tests/qsort_external_variable_err.out
@@ -1,7 +1,7 @@
-vlib/v/checker/tests/qsort_external_variable_err.vv:8:22: error: can not access external variable `i`
-    6 |     unsafe{array.sort(a[0]>b[0])}
-    7 |     println(array)
-    8 |     unsafe{array.sort(a[i]<b[i])}
-      |                         ^
-    9 |     println(array)
-   10 | }
+vlib/v/checker/tests/qsort_external_variable_err.vv:7:24: error: can not access external variable `i`
+    5 |     unsafe { array.sort(a[0] > b[0]) }
+    6 |     println(array)
+    7 |     unsafe { array.sort(a[i] < b[i]) }
+      |                           ^
+    8 |     println(array)
+    9 | }

--- a/vlib/v/checker/tests/qsort_external_variable_err.out
+++ b/vlib/v/checker/tests/qsort_external_variable_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/qsort_external_variable_err.vv:8:22: error: can not access external variable `i`
+    6 |     unsafe{array.sort(a[0]>b[0])}
+    7 |     println(array)
+    8 |     unsafe{array.sort(a[i]<b[i])}
+      |                         ^
+    9 |     println(array)
+   10 | }

--- a/vlib/v/checker/tests/qsort_external_variable_err.vv
+++ b/vlib/v/checker/tests/qsort_external_variable_err.vv
@@ -1,0 +1,9 @@
+fn main() {
+	array := [[1, 5], [4, 2]]
+	println(array)
+	i := 0
+	unsafe { array.sort(a[0] > b[0]) }
+	println(array)
+	unsafe { array.sort(a[i] < b[i]) }
+	println(array)
+}

